File: 剑指offer/28.对称的二叉树.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        def help(left, right):
            if not left and not right:
                return True

            if not left or not right or left.val != right.val:
                return False

            return help(left.left, right.right) and help(
                left.right, right.left)

        if not root:
            return True

        return help(root.left, root.right)

File: 剑指offer/28.对称的二叉树.cpp
#include <iostream>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};

class Solution {
public:
    bool dfs(TreeNode* p, TreeNode* q)
    {
        if (!q || !p)
            return !p && !q;
        if (p->val != q->val)
            return false;
        return dfs(p->left, q->right) && dfs(p->right, q->left);
    }

    bool isSymmetric(TreeNode* root)
    {
        if (!root)
            return true;
        return dfs(root->left, root->right);
    }
};